### ISSUE TITLE
Upgrade to libjpeg-turbo

### ISSUE THREAD

**Issue by mkolod** [tensorflow/tensorflow#4807]
TensorFlow seems to be using the classic libjpeg v 0.9, but libjpeg-turbo is heavily optimized to leverage SIMD on various platforms (x86, ARM, POWER, etc.), and is both API and ABI compatible with libjpeg, at least up to libjpeg 0.8. Given TensorFlow's goal of portability across a wide varierty of platforms, I'm wondering if libjpeg was decided upon instead of libjpeg-turbo for portability, or was this just an omission? 

libjpeg-turbo speeds up JPEG decoding by about 2x, and since that reduces decoding latency and increases throughput, it not only improves performance of TensorFlow on the CPU, but especially on the GPU, which tends to wait for the CPU to provide decoded images. Moreover, that could result in a major performance improvement on more CPU-underpowered platforms such as ARM (e.g. via NEON SIMD on the NVIDIA TX1 platform which uses the NEON-capable Cortex A57, or on ARM Cortex A8 on a Samsung Galaxy S).

**Comment by tatatodd** [comment]
@mkolod Thanks for filing this issue!

Switching to libjpeg-turbo sounds like a reasonable thing to investigate.  We welcome contributions!

If anyone is going to attempt this, probably the two areas to focus on are:
1. Benchmarks showing the actual performance when used within TensorFlow.
2. A breakdown of the portability tradeoffs between the two libraries.

Adding @prb12 @poxvoculi since they might be interested, and might know some of the history behind why libjpeg was chosen over libjpeg-turbo (if any).

**Comment by jhseu** [comment]
We should definitely switch. libjpeg-turbo supports all the platforms we're on, and it seems that both Chrome and Firefox are already using it.

**Comment by jhseu** [comment]
Self-assigning so I can test it out internally as well.

**Comment by mkolod** [comment]
@jhseu I actually managed to add it in my own experiments, but it seems to be failing at runtime, even though compilation succeeds (more on this in a moment). I suppose the big question about how to distribute such a change to the community is how to deal with the fact that the libjpeg-turbo code is largely assembly. When building from source, it may be tricky doing it from Bazel without say genrules. Also, the current assembly mnemonics specifically call for nasm, though maybe there's an easier way. In my experiments, I have a genrule that calls turbo's own build, which is based on autotools (autoconf/configure/make, with make calling nasm), but that's a Rube Goldberg solution at best. Another option is of course to pull a pre-built static lib (libjpeg.a) from some CDN, removing the dependence on dealing with building assembly code, and just having a cc_library that provides headers and uses a pre-built lib. I'm not sure which approach would be preferable to the community.

**Comment by mkolod** [comment]
@jhseu Regarding the runtime failure after swapping out the libjpeg v9-based libjpeg.a with libjpeg-turbo based lib (API and ABI-compatible with libjpeg v8), I swapped out both the static lib (compiled with -fPIC) and the headers, the Bazel build completed on my machine, but there's a runtime failure while decoding. There's no clear exception, the queues that run the jpeg decoding just shut down with an OutOfRangeError (must be a silent error). I'm planning to start debugging that tomorrow. Generally, libjpeg-turbo with API and ABI support for libjpeg v8 should be good enough, because libjpeg v9 only added a nonstandard color transform to "Lossless SmartScale," which is neither a community nor industry standard (only supported in libjpeg v9 and no other library or image capture device). See [here](http://www.libjpeg-turbo.org/About/Jpeg-9) about why libjpeg-turbo only goes until libjpeg v8 support. So, it's a bit odd that there's an issue since unless there's an API incompatibility, libjpeg-turbo is a drop-in for libjpeg. I'll keep you posted what I find and then share a branch on a fork or something. Like I said, my solution may not be ideal, it's tricky dealing with assembly-based builds, and it's tricker still to know what's preferable from the project's overall standards point of view for custom code like this.

**Comment by mkolod** [comment]
@jhseu BTW, the way Caffe deals with turbo is that it just uses OpenCV, which comes with it. I'm not sure about whether other frameworks upgraded to turbo (Torch hasn't, I know that almost for sure), but that assembly code base makes for a complex build for sure.

**Comment by mkolod** [comment]
@jhseu Also, many Linux distros already come with libjpeg-turbo, but I assume that relying on the uncertainty of whether a shared lib is there is complicated if one is to support platforms other than Linux. So, some static linking would probably be better, which is how libjpeg was already integrated (but libjpeg was pure C so it could use a cc_library rule).

**Comment by mkolod** [comment]
@jhseu Also for a quick benchmark, I've done AlexNet profiling, and JPEG decode was consuming about 30% of the CPU time, with Core i7 6 cores fully utilized serving one GTX 1080. AlexNet is a good case here because it's light on GPU compute and exposes inefficiencies elsewhere in the system, such as I/O (frequent gradient updates, lots of disk requests for large datasets read via queues, e.g. ImageNet), etc. JPEG decoding is important here because the GPUs tend to wait for the CPU to finish decoding and so the GPU utilization is low if decoding takes too long. I found that AlexNet training (batch size 128-1,024) could be sped up by 20% on average on above mentioned GTX 1080. This may matter less for say Inception v3 or ResNet-50, but it will still help, particularly in multi-GPU environments which are more CPU-bound for decoding.

**Comment by jhseu** [comment]
@mkolod Ah, thanks for researching this so thoroughly.

As a general principle, we prefer to build and statically link dependencies for all required features. This simplifies deployment in certain settings (e.g., distributed training) and helps with reproducibility.

The right thing to do is to create a new rule that underneath uses genrule to compile assembly with nasm. I'm unsure about whether we'd want to add nasm as a required build-time dependency, though.

Unassigning since it's not clear we want to make libjpeg-turbo default because of the nasm dependency. Making it optional seems reasonable, though.

**Comment by mkolod** [comment]
@jhseu I'll try to look into making turbo optional in the Bazel build, right now I have it on for all cases. Meanwhile, I wanted to share some perf figures of how this would impact decoding. The base case here is Caffe and CNTK, which use OpenCV and get about 1k ImageNet images decoded per CPU core on a CPU the same as mine (6-core Core i7-5930K 3.50GHz). I used the standard ImageNet protobufs generated using the ImageNet preprocessing script from the Inception code base. The figures are as follows, on a per core basis:

base case: 219 images/sec

base + Fast IDCT (see issue #5072): 233 im/sec

base + Fast IDCT + libjpeg-turbo: 356 im/sec

base + Fast IDCT + libjpeg-turbo + pre-resized images (to 256x256, can crop it down to 227x227 for AlexNet at runtime with random crops, etc.): 700 im/sec

The Caffe/CNTK perf is at 1k im/sec, but that's also in benchmarks with pre-resized images (one shouldn't have to do anything else than cropping at runtime, definitely not padding, etc.). For lighter networks like AlexNet, the image decode is definitely a bottleneck. For heavier networks like Inception, it may not be an issue, but again, it could be an issue for more CPU-challenged platforms like ARM.

[Here](https://github.com/mkolod/tf_perf_eval/blob/master/io/only_decode_benchmark.py) is the script I used for this benchmark. I'll post the Bazel integration with the libjpeg-turbo build shortly. For making turbo optional, I wonder if it shouldn't be a configure option, like the CUDA/CUDNN/HDFS support right now? If not, what's another option for conditional inclusion into the build?

**Comment by jhseu** [comment]
Yeah, as a ./configure option makes sense. We also optionally allow configuration through environment variables with running ./configure, so it can be a hidden option.

We should check for the presence of nasm during ./configure if the user enables it.

**Comment by mkolod** [comment]
@jhseu Yes, I'm following the HDFS support, in that I provide options for both interactive selection and the env var. Furthermore, invalid selections, such as choosing libjpeg-turbo when the build support for assembly is not available, causes a configure failure and a recommendation to install the appropriate tools, or select N during the libjpeg-turbo prompt (or selecting 0 using the env var) to fall back to libjpeg. This should make libjpeg-turbo completely optional, and improve performance for users who have the appropriate tooling on their OS or Docker container.

**Comment by jart** [comment]
Hey @mkolod did https://github.com/tensorflow/tensorflow/commit/52ffc792fdef41de20ac60dc61d4e809d83b558d end up working out for you? It would be helpful to get feedback. Especially if you have time to review the BUILD file, in case there's some select() statements we forgot to write tuning those constants for various platforms.
